metals/src/main/scala/scala/meta/internal/metals/codeactions/FlatMapToForComprehensionCodeAction.scala

```diff
@@ -0,0 +1,541 @@
+          .findLastEnclosingAt[Term.Apply](
+            path,
+            range.getStart(),
+            isTreeInteresting
+          )
+      else
+        None
+
+    val maybeChainedCodeAction = for {
+      document <- buffers.get(path)
+      termApply <- maybeTree
+      termApplyArg <- termApply.args.headOption
+      indentation = getIndentForPos(termApplyArg.pos, document)
+    } yield codeActionWithApply(
+      path,
+      termApply,
+      indentation
+    )
+    maybeChainedCodeAction.flatten.toSeq
+
+  }
+
+  private def constructCodeAction(
+      forElementsList: List[Enumerator],
+      yieldTerm: Term,
+      indentation: String,
+      path: AbsolutePath,
+      startPos: l.Position,
+      endPos: l.Position
+  ): l.CodeAction = {
+
+    val indentedElems = forElementsList
+      .map(
+        _.syntax
+          .split(Array('\n'))
+          .map(line => s"$indentation   $line")
+          .mkString("\n")
+      )
+      .mkString("\n")
+
+    val yieldTermIndentedString = yieldTerm.syntax
+      .split(Array('\n'))
+      .map(line => s"$indentation   $line")
+      .mkString("\n")
+
+    val forYieldString =
+      s"""|{
+          |$indentation for {
+          |$indentedElems
+          |$indentation }  yield {
+          |$yieldTermIndentedString
+          |$indentation }
+          |$indentation}""".stripMargin
+
+    val codeAction = new l.CodeAction()
+    val range =
+      new l.Range(startPos, endPos)
+    codeAction.setTitle(
+      FlatMapToForComprehensionCodeAction.flatMapToForComprehension
+    )
+    codeAction.setKind(this.kind)
+    val forComprehensionTextEdit = new l.TextEdit(range, forYieldString)
+    codeAction.setEdit(
+      new l.WorkspaceEdit(
+        Map(
+          path.toURI.toString -> List(forComprehensionTextEdit).asJava
+        ).asJava
+      )
+    )
+    codeAction
+  }
+
+  private def codeActionWithApply(
+      path: AbsolutePath,
+      termApply: Term.Apply,
+      indentation: String
+  ): Option[l.CodeAction] = {
+    val (forElements, maybeYieldTerm, _) = {
+      extractChainedForYield(
+        None,
+        None,
+        List.empty,
+        termApply,
+        Set.empty
+      )
+    }
+
+    if (forElements.nonEmpty) {
+      maybeYieldTerm.map { yieldTerm =>
+        constructCodeAction(
+          forElements,
+          yieldTerm,
+          indentation,
+          path,
+          termApply.pos.toLSP.getStart,
+          termApply.pos.toLSP.getEnd
+        )
+      }
+    } else None
+
+  }
+
+  private def replacePlaceHolder(
+      tree: Term,
+      newName: Term.Name,
+      allowedToGetInsideApply: Boolean
+  ): (Term, Int) =
+    tree match {
+      case Term.Apply(fun, args) if allowedToGetInsideApply =>
+        val (newFun, funReplacementTimes) =
+          replacePlaceHolder(fun, newName, false)
+        val (newArgs, argsReplacementTimes) =
+          args.map(replacePlaceHolder(_, newName, false)).unzip
+
+        val replacementTimes =
```

Comment:
I would say, let’s not show any code action if the user has non-compiling code, because that non-compiling code, might be a draft, or the user might actually have made some mistake, which they are themselves supposed to be correcting. I think showing a code action to fix non-compiling code, would cover the problems for especially non-experienced users. So someone might be processing a list of tuples, and might expect that the map should consume a two parameter function, so the second placeholder might in the mind of the programmer be assumed to be the right side value of the tuples on the list. let’s leave it to users.

Reply:
This is only a syntactic conversion, so I am not sure if we should not show action if there is a compilation error. Users might want to do a conversion on a non compiling code and fix it later, since they are probably editing things anyway.

Reply:
but, we are hiding the problem away, by doing a conversion and replacing arguments.

Reply:
> but, we are hiding the problem away, by doing a conversion and replacing arguments.

Ok, we can try to be more correct in here, let's just make sure that we don't suggested an code action in that case.

Reply:
Did we add this test? It should be fine to add and just use the right flag.